File: packages/sdk-flutter/lib/src/media.dart
import 'client.dart';
import 'http.dart';

/// One item in the project's media library, as returned by the API.
class MediaItem {
  final int id;
  final String fileName;
  final String fullUrl;
  final String thumbUrl;
  final String? caption;
  final int size;
  final int? width;
  final int? height;

  const MediaItem({
    required this.id,
    required this.fileName,
    required this.fullUrl,
    required this.thumbUrl,
    required this.size,
    this.caption,
    this.width,
    this.height,
  });

  factory MediaItem.fromJson(Map<String, dynamic> json) => MediaItem(
        id: json['id'] as int,
        fileName: json['fileName'] as String,
        fullUrl: json['fullUrl'] as String,
        thumbUrl: json['thumbUrl'] as String,
        caption: json['caption'] as String?,
        size: json['size'] as int,
        width: json['width'] as int?,
        height: json['height'] as int?,
      );
}

/// A page of [MediaItem]s, as returned by [MediaResource.list].
class MediaPage {
  final List<MediaItem> data;
  final int? total;
  final int? page;

  const MediaPage({required this.data, this.total, this.page});

  factory MediaPage.fromJson(Map<String, dynamic> json) => MediaPage(
        data: (json['data'] as List)
            .map((e) => MediaItem.fromJson(e as Map<String, dynamic>))
            .toList(),
        total: json['total'] as int?,
        page: json['page'] as int?,
      );
}

/// A file to upload -- raw bytes plus the filename/content-type the
/// server should store it as. Build this from `File.readAsBytes()`
/// (dart:io, non-web), an `XFile`/`image_picker` result's `.readAsBytes()`,
/// or a `<input type=file>` pick on Flutter web -- this package doesn't
/// depend on any of those, so wiring up the actual file picker is your
/// app's job.
class UploadableFile {
  final List<int> bytes;
  final String filename;

  /// e.g. `'image/png'`. Images are thumbnailed server-side automatically;
  /// other file types are stored as-is.
  final String? contentType;

  const UploadableFile({required this.bytes, required this.filename, this.contentType});
}

/// This project's media library (list/upload/updateCaption/delete).
class MediaResource {
  final VeroXMApp _app;

  MediaResource(this._app);

  Future<MediaPage> list({int? page, String? search}) async {
    final result = await _app.request(RawRequestOptions(
      path: '/media',
      query: {'page': page?.toString(), 'search': search},
    ));
    return MediaPage.fromJson(result as Map<String, dynamic>);
  }

  /// Max size is whatever this project's server allows (commonly 20MB).
  Future<MediaItem> upload(UploadableFile file, {String? caption}) async {
    final result = await _app.request(RawRequestOptions(
      method: 'POST',
      path: '/media/upload',
      multipart: MultipartUpload(
        bytes: file.bytes,
        filename: file.filename,
        contentType: file.contentType,
        caption: caption,
      ),
    ));
    return MediaItem.fromJson(result as Map<String, dynamic>);
  }

  Future<MediaItem> updateCaption(int id, String caption) async {
    final result = await _app.request(RawRequestOptions(
      method: 'PATCH',
      path: '/media/$id',
      json: {'caption': caption},
    ));
    return MediaItem.fromJson(result as Map<String, dynamic>);
  }

  Future<void> delete(int id) async {
    await _app.request(RawRequestOptions(method: 'DELETE', path: '/media/$id'));
  }
}

File: packages/sdk-js/src/media.ts
import type { VeroXMApp } from './client.js';

export interface MediaItem {
  id: number;
  fileName: string;
  fullUrl: string;
  thumbUrl: string | null;
  caption: string | null;
  size: number | null;
  width: number | null;
  height: number | null;
}

export interface MediaPage {
  data: MediaItem[];
  page: number;
  perPage: number;
  total: number;
}

/**
 * A Blob/File (browser -- e.g. straight from an `<input type="file">` or
 * `fetch()` response) uploads as-is; the `{data, filename}` form covers
 * Node/React Native, where raw bytes usually arrive as an ArrayBuffer or
 * Uint8Array rather than a Blob.
 */
export type UploadableFile = Blob | { data: Blob | ArrayBuffer | Uint8Array; filename: string; contentType?: string };

export class MediaResource {
  constructor(private readonly app: VeroXMApp) {}

  async list(options: { page?: number; search?: string } = {}): Promise<MediaPage> {
    return this.app.request({
      path: '/media',
      query: { page: options.page, search: options.search },
    }) as Promise<MediaPage>;
  }

  /** Images are thumbnailed server-side automatically; other file types are stored as-is. Max size is whatever this project's server allows (commonly 20MB). */
  async upload(file: UploadableFile, caption?: string): Promise<MediaItem> {
    const form = new FormData();
    if (file instanceof Blob) {
      form.append('file', file);
    } else {
      // TS's DOM lib types Uint8Array as generic over ArrayBufferLike (which
      // includes SharedArrayBuffer), which BlobPart's own typing doesn't
      // accept -- a type-checker-only mismatch, not a real runtime one,
      // since both an ArrayBuffer and any ArrayBufferView are valid Blob
      // constructor parts in every actual runtime this SDK targets.
      const blob = file.data instanceof Blob ? file.data : new Blob([file.data as BlobPart], { type: file.contentType });
      form.append('file', blob, file.filename);
    }
    if (caption) form.append('caption', caption);

    return this.app.request({ method: 'POST', path: '/media/upload', form }) as Promise<MediaItem>;
  }

  async updateCaption(id: number, caption: string): Promise<MediaItem> {
    return this.app.request({
      method: 'PATCH',
      path: `/media/${id}`,
      json: { caption },
    }) as Promise<MediaItem>;
  }

  async delete(id: number): Promise<void> {
    await this.app.request({ method: 'DELETE', path: `/media/${id}` });
  }
}
